common/http/platform/beast/http.cpp

```diff
@@ -0,0 +1,344 @@
+		string ips = "[";
+		string sep;
+		for (auto r : results) {
+			ips += sep;
+			ips += r.endpoint().address().to_string();
+			sep = ", ";
+		}
+		ips += "]";
+		logger_.Debug("Hostname " + request_->host_ + " resolved to " + ips);
+	}
+
+	resolver_results_ = results;
+
+	this->stream_.async_connect(
+		resolver_results_, [this](error_code err, const asio::ip::tcp::endpoint &endpoint) {
+			ConnectHandler(err, endpoint);
+		});
+}
+
+void HttpSession::ConnectHandler(error_code err, const asio::ip::tcp::endpoint &endpoint) {
+	if (err) {
+		CallErrorHandler(err, request_, header_handler_);
+		return;
+	}
+
+	logger_.Debug("Connected to " + endpoint.address().to_string());
+
+	http_request_ = make_shared<http::request<http::buffer_body>>(
+		VerbToBeastVerb(request_->method_), request_->path_, BeastHttpVersion);
+	http_request_serializer_ =
+		make_shared<http::request_serializer<http::buffer_body>>(*http_request_);
+
+	http::async_write_header(
+		stream_, *http_request_serializer_, [this](error_code err, size_t num_written) {
+			WriteHeaderHandler(err, num_written);
+		});
+}
+
+void HttpSession::WriteHeaderHandler(error_code err, size_t num_written) {
+	if (err) {
+		CallErrorHandler(err, request_, header_handler_);
+		return;
+	}
+
+	logger_.Trace("Wrote " + to_string(num_written) + " bytes of header data to stream.");
+
+	auto header = request_->GetHeader("Content-Length");
+	if (!header || header.value() == "0") {
+		ReadHeader();
+		return;
+	}
+
+	auto length = common::StringToLong(header.value());
+	if (!length || length.value() < 0) {
+		auto err = error::Error(
+			length.error().code, "Content-Length contains invalid number: " + header.value());
+		CallErrorHandler(err, request_, header_handler_);
+		return;
+	}
+	request_body_length_ = length.value();
+
+	if (!request_->body_gen_) {
+		auto err = MakeError(BodyMissingError, "Content-Length is non-zero, but body is missing");
+		CallErrorHandler(err, request_, header_handler_);
+		return;
+	}
+
+	auto body_reader = request_->body_gen_();
+	if (!body_reader) {
+		CallErrorHandler(body_reader.error(), request_, header_handler_);
+		return;
+	}
+	request_->body_reader_ = body_reader.value();
+
+	WriteBody();
+}
+
+void HttpSession::WriteBodyHandler(error_code err, size_t num_written) {
+	if (err) {
+		CallErrorHandler(err, request_, header_handler_);
+		return;
+	}
+
+	logger_.Trace("Wrote " + to_string(num_written) + " bytes of body data to stream.");
+
+	if (num_written == 0) {
+		// We are ready to receive the response.
+		ReadHeader();
+	} else {
+		// We are still writing the body.
+		WriteBody();
+	}
+}
+
+void HttpSession::WriteBody() {
+	auto read = request_->body_reader_->Read(body_buffer_.begin(), body_buffer_.end());
+	if (!read) {
+		CallErrorHandler(read.error(), request_, header_handler_);
+		return;
+	}
+
+	http_request_->body().data = body_buffer_.data();
+	http_request_->body().size = read.value();
+
+	if (read.value() > 0) {
+		http_request_->body().more = true;
```

Comment:
So we always call this at least twice if there is some body data because we can't say if everything was read until `Read()` returns `0`?

Reply:
Yes, I pondered on this one for a while. But I don't think there's a way. We cannot even attempt to read because as soon as we do we might end up with more data than we can handle.